z/btree.go

```diff
@@ -65,24 +68,84 @@ func NewTree(tag string) *Tree {
 
 // NewTree returns a persistent on-disk B+ tree.
 func NewTreePersistent(path string) (*Tree, error) {
-	buffer, err := NewBufferPersistent(path)
+	t := &Tree{}
+	var err error
+
+	// Open the buffer from disk and set it to the maximum allocated size.
+	t.buffer, err = NewBufferPersistent(path, minSize)
 	if err != nil {
 		return nil, err
 	}
-	tree := &Tree{buffer: buffer}
-	tree.Reset()
-	return tree, nil
+	t.buffer.offset = uint64(len(t.buffer.buf))
+	t.data = t.buffer.Bytes()
+
+	// TODO(ajeet): can we do better than this?
+	isInitialized := false
+	for _, b := range t.data {
+		if b != 0 {
+			isInitialized = true
+			break
+		}
+	}
+	if !isInitialized {
+		t.nextPage = 1
+		t.freePage = 0
+		return t, nil
+	}
+
+	// Calculate t.nextPage by finding the highest pageId among all the nodes.
+	t.nextPage = 0
+	t.Iterate(func(n node) {
+		if pageId := n.pageID(); pageId > t.nextPage {
+			t.nextPage = pageId
+		}
+	})
+
+	// Calculate t.freePage by finding the page to which no other page points.
+	// usedPages[i] is true if pageId i+1 is in use.
+	usedPages := make([]bool, t.nextPage)
+	// If a node points to a page, mark it as used.
+	t.Iterate(func(n node) {
+		i := n.pageID() - 1
+		usedPages[i] = true
+	})
+	// pointPages is a list of page IDs that the free pages point to.
+	pointedPages := make([]uint64, 0)
+	for i, used := range usedPages {
+		if !used {
+			pageId := uint64(i) + 1
+			pointedPages = append(pointedPages, t.node(pageId).uint64(0))
+		}
+	}
+	// Mark all pages being pointed to as used.
+	for _, pageId := range pointedPages {
+		i := pageId - 1
+		usedPages[i] = true
+	}
+	// Only one page would still be marked as free.
+	// This is the root free page node.
+	for i, used := range usedPages {
```

Comment:
`usedPages` is slightly confusing name. Can we name it better?

Reply:
Hm, maybe. What should I name it to? I could name it `freePages` and flip all the bits, but that seems unnecessary.

Reply:
Renamed to `childPages`.